### ISSUE TITLE
Deprecate pickling support (__getstate__ and __setstate__) in twisted.internet._sslverfiy

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6166]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#6166|
|Type|enhancement|
|Created|2012-10-29 15:01:14Z|
|Branch|https://github.com/twisted/twisted/tree/sslverify-no-pickle-6166-4|

Pickleability is no longer a goal for Twisted. As such, the various `__getstate__` and `__setstate__` methods of `twisted.internet._sslverify` classes should be deprecated, so we can remove the extra code involved.

Attachments:

* [ticket6166.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/490/490b33c920eb6cd62734d2d4b06dcac0bf62ce66/c5d956315a0e10ea053e0df2eaca3bc2cc1ee2e5.patch) (12379 bytes) - added by canibanoglu on 2013-04-15 20:54:17Z - Deprecation of __getstate__ and __setstate__ along with new test which check for deprecation
* [ticket#6166_2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/490/490b33c920eb6cd62734d2d4b06dcac0bf62ce66/f82c55803da8a9cf91ceb99e040adc798fc07d44.patch) (9529 bytes) - added by canibanoglu on 2013-04-17 11:27:24Z - New patch on trunk which deprecates __getstate__ and __setstate in twisted.internet._sslverify and adds a 6166.removal in twisted/topfiles
* [ticket6166_2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/490/490b33c920eb6cd62734d2d4b06dcac0bf62ce66/46978a360853cae4a659dc2a522336f2a0d19094.patch) (9529 bytes) - added by canibanoglu on 2013-04-17 15:05:57Z - Removed # from the filename
* [ticket6166-3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/490/490b33c920eb6cd62734d2d4b06dcac0bf62ce66/849b922ffecb91640ab320a390f6aa1fb1b9b3a3.patch) (5523 bytes) - added by canibanoglu on 2013-04-18 21:26:11Z - Tests now use callDeprecated to test deprecations, use .suppress attribute to suppress deprecation warning messages

<details><summary>Searchable metadata</summary>

```
trac-id__6166 6166
type__enhancement enhancement
reporter__itamar itamar
priority__low low
milestone__ 
branch__branches_sslverify_no_pickle_6166_4 branches/sslverify-no-pickle-6166-4
branch_author__canibanoglu__tomprince canibanoglu, tomprince
status__closed closed
resolution__fixed fixed
component__core core
keywords__easy_gsoc easy gsoc
time__1351522874000000 1351522874000000
changetime__1397660429000000 1397660429000000
version__None None
owner__tom_prince tom.prince
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst commented|
|-|-|

See `twisted/python/reflect.py` and `twisted/test/test_reflect.py` for an example of deprecation and testing of deprecation.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_4"></a>|@tomprince commented|
|-|-|

The patch you attached is reversed, and appears to be against the most recent release, rather than trunk. If you are planning on contributing to twisted, you should be working in a source checkout (see [TwistedDevelopment]).

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_5"></a>|@tomprince set owner to canibanoglu|
|-|-|

The attached patch appears to not apply to trunk.

 1. The deprecations want to use `twisted.python.deprecate.deprecated`, rather than calling `warnings.warn` directly. The message should probably say something about pickling not being supported anymore.
 2. Running
```
python -Wall bin/trial twisted.test.test_sslverify
```
  gives me a bunch of deprecation warnings. They should be suppressed.

**Comment by twisted-trac** [comment]
|<img alt="canibanoglu's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|canibanoglu commented|
|-|-|

Thanks a lot for the input.

First of all, the release and trunk version mixup was bordering on idiocy, sorry about that. 

1. I have tried using twisted.python.deprecate.deprecated but it doesn't pass the tests because it doesn't work the same way for class methods and module-level functions. I found a ticket which is now marked as fixed but deprecated still acts in a peculiar way. More specifically this is what I get

```twisted.trial.unittest.FailTest: not equal:
a = 'twisted.internet._sslverify.OpenSSLCertificateOptions.__getstate__ was deprecated in Twisted 13.1.0'
b = 'twisted.internet._sslverify.__getstate__ was deprecated in Twisted 13.1.0'```

2. Thank you very much for pointing that out, I will definitely take care of that.

**Comment by twisted-trac** [comment]
|<img alt="canibanoglu's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|canibanoglu removed owner|
|-|-|

Here's my second go at it. 

- I have suppressed deprecation warnings when running the tests with 
```
python -Wall bin/trial twisted.test.test_sslverify
```
- Still can't get @deprecated to work correctly with class methods. This patch still uses warnings.warn for deprecation warnings. Added "Pickling is no longer supported." to the deprecation warning message.

- Added 6166.removal topfile.

I have a question though. While working on this I have found that @deprecated decorator is not functioning as expected when decorating class methods and the full test run hangs on OS X Mountain Lion (10.8). There are tickets for both of these and they both appear as fixed.  Should I create another ticket?

Thanks a lot for the assistance.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

> New patch on trunk which ...

Can you re-attach the patch, this time without a space in the filename?  trac cannot serve attachments with spaces in them.

**Comment by twisted-trac** [comment]
|<img alt="canibanoglu's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|canibanoglu commented|
|-|-|

There's an underscore between 6166 and 2 :) Should I rename them differently from now on?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

Sorry.  I guess there's also a problem with filenames with "#" in them.

**Comment by twisted-trac** [comment]
|<img alt="canibanoglu's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|canibanoglu commented|
|-|-|

Uploaded the same patch file without the "#" character in the filename.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_12"></a>|@tomprince set owner to canibanoglu|
|-|-|

1.There are heplers to suppres warnings. See `twisted.trial.util.suppress` and the `.suppress` attribute of `TestCase` and/or `test_*` methods.
 2. Hmm. Looking at the r24931 diff for [#3254](https://github.com/twisted/twisted/issues/3254), `deprecated` works with methods, but it needs to called as
 ```
 Class.method = deprecated(version)(Class.method)
 ```
 after the class definition. You should file a ticket to support the obvious thing.
   - (Making this work is probably tricky, and would involve making `deprecate` return a descriptor)
 3. Is there any reason you created your own helper, rather than using `callDeprecated`?
 4. The newsfile could be improved by listing all the things that are getting deprecated, rather than saying each individual thing is deprecated, but might be better to just say that pickling the two relevant classes is now deprecated.
 5. Even if not using `@deprecated`, you should still use the same format for the docstring.

**Comment by twisted-trac** [comment]
|<img alt="canibanoglu's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|canibanoglu removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_14"></a>|@tomprince commented|
|-|-|

(In [38505]) Branching to sslverify-no-pickle-6166.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_15"></a>|@tomprince set owner to @tomprince<br>@tomprince set status to `assigned`|
|-|-|

This looks good. A couple of points:
 1. The suppression for `test_certificateOptionsSerialization` is to general.
 2. `CertificateOptions` and `KeyPair` both live publicly in `twisted.inte
rnet.ssl`, so the news should reflect that.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_16"></a>|@tomprince commented|
|-|-|

(In [38507]) Address review comments.

Refs [#6166](https://github.com/twisted/twisted/issues/6166).

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_17"></a>|@tomprince commented|
|-|-|

(In [38518]) Be even more specific about what to suppress.

Refs [#6166](https://github.com/twisted/twisted/issues/6166).

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_18"></a>|@tomprince commented|
|-|-|

To expand one (1), we only want to suppress the specific warnings that we are triggering by testing this code. That way, if somebody later changes something else to cause a different warning to occur here, it would still be reported.

Other than those two points, this looks good.

I'll go ahead and merge after fixing those two issues.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_19"></a>|@tomprince set status to `closed`|
|-|-|

(In [38519]) Merge sslverify-no-pickle-6166: Deprecate pickling support in twisted.internet._sslverfiy.

Authore: canibanoglu
Reviewers: tom.prince
Fixes: [#6166](https://github.com/twisted/twisted/issues/6166)

Pickleability is no longer a goal for Twisted. As such, the
various __getstate__ and __setstate__ methods of
twisted.internet._sslverify classes can be deprecated, so we
can remove the extra code involved.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun set status to `reopened`|
|-|-|

(In [38527]) Revert r38517 - test suite regression

Reopens: [#6166](https://github.com/twisted/twisted/issues/6166)

A number of test modules now fail to import on some builders:

```
[ERROR]
Traceback (most recent call last):
  File "/data/home/buildbot/work/freebsd-8.2-i386/Twisted/twisted/trial/runner.py", line 500, in loadPackage
    module = modinfo.load()
  File "/data/home/buildbot/work/freebsd-8.2-i386/Twisted/twisted/python/modules.py", line 383, in load
    return self.pathEntry.pythonPath.moduleLoader(self.name)
  File "/data/home/buildbot/work/freebsd-8.2-i386/Twisted/twisted/python/_reflectpy3.py", line 266, in namedAny
    topLevelPackage = _importAndCheckStack(trialname)
  File "/data/home/buildbot/work/freebsd-8.2-i386/Twisted/twisted/python/_reflectpy3.py", line 205, in _importAndCheckStack
    return __import__(importName)
  File "/data/home/buildbot/work/freebsd-8.2-i386/Twisted/twisted/test/test_sslverify.py", line 805, in <module>
    class KeyPair(unittest.TestCase):
  File "/data/home/buildbot/work/freebsd-8.2-i386/Twisted/twisted/test/test_sslverify.py", line 808, in KeyPair
    CN=b"server")[0]
  File "/data/home/buildbot/work/freebsd-8.2-i386/Twisted/twisted/test/test_sslverify.py", line 84, in makeCertificate
    keypair = PKey()
exceptions.NameError: global name 'PKey' is not defined

twisted.test.test_sslverify
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_21"></a>|@exarkun commented|
|-|-|

(In [38528]) whitespace fixes

refs [#6166](https://github.com/twisted/twisted/issues/6166)

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_22"></a>|@tomprince commented|
|-|-|

(In [38537]) Branching to sslverify-no-pickle-6166-2.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_23"></a>|@tomprince commented|
|-|-|

I think that should address the test failures.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)|@tomprince removed owner<br>@tomprince set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_25"></a>|@glyph commented|
|-|-|

Minor nitpick: [there is a new twisted checker error](http://buildbot.twistedmatrix.com/builders/twistedchecker/builds/709/steps/run-twistedchecker/logs/new%20twistedchecker%20errors); a new line is too long.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_26"></a>|@tomprince commented|
|-|-|

(In [38926]) Wrap long line.

Refs: [#6166](https://github.com/twisted/twisted/issues/6166)

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_27"></a>|@habnabit commented|
|-|-|

There was no buildbot run for this, so I'm forcing one now.

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)|@habnabit set owner to @habnabit|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)|@wallrj set owner to @wallrj<br>@wallrj set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_30"></a>|@wallrj set owner to @tomprince<br>@wallrj set status to `new`|
|-|-|

Notes:
 * Merges cleanly

Points:
 1. Build Failures
    1. Python3 tests are very unhappy: https://buildbot.twistedmatrix.com/builders/python-3.3-tests/builds/1405/steps/shell/logs/stdio
    2. The circularChainWarning failure is something that has since
       been fixed in trunk I think (r39074)
    3. The output of python3 tests seems unusual. It only mentions
       failed and skipped tests, not the ones that passed.
    4. It also says it dropped to a debugger which I don't remember
       seeing before.
    5. So try merging forward? Maybe that'll fix all the skips too.
 2. Deprecation version numbers need updating. 13.1.0 > 13.2.0
 3. [branches/sslverify-no-pickle-6166-2/twisted/test/test_sslverify.py](branches/sslverify-no-pickle-6166-2/twisted/test/test_sslverify.py)
    1. "util.suppress(category = DeprecationWarning" -- remove
       whitespace in the argument assignment.
    2. "Test deprecation" -- remove the "Test" from test docstrings.
    3. "test_setstateDeprecation.suppress" --
       1. Do you have to pass a full state dictionary in the
          __setstat__ test?
       2. If you could just pass an empty dict (or minimal dummy
          dict) this extra suppression might not be necessary.
       3. The wrapping and message formatting looks strange -- don't
          the quoted strings need to be surrounded in round brackets?
       4. Fix the whitespace in the argument assignment.
    4. "KeyPair" many of the same comments apply to this new testcase.
 4. [branches/sslverify-no-pickle-6166-2/twisted/internet/_sslverify.py](branches/sslverify-no-pickle-6166-2/twisted/internet/_sslverify.py)
    1. Can you move the deprecation decorators directly beneath the
       affected function, to make it obvious that it's deprecated.

Please, merge forward, address or answer the numbered points above and
link to clean build results then merge.

Thanks.

-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_31"></a>|@tomprince commented|
|-|-|

(In [42225]) Branching to sslverify-no-pickle-6166-3.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_32"></a>|@tomprince commented|
|-|-|

The decorators have to be after the class definition, unfortunately.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)|@tomprince removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_34"></a>|@exarkun set owner to @tomprince|
|-|-|

Thanks.

  1. There's a twistedchecker error in `_sslverify.py` that it might be nice to fix here.
  1. The suppression is defined as a regular expression.  This code takes advantage of that fact in some ways but ignores it in others - for example, it doesn't escape the "." module separator.
  1. It's unfortunate that 14.0 hasn't been released yet.  **I** think that 14.1 is the correct version for this deprecation.  It wouldn't surprise me if some people try to restart the 14.0 release process based on a new trunk revision though, that would cause a problem here.  I don't know what to do about that except blame those people for screwing things up if that happens, though.

Please address those points and then merge.  Thanks.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_35"></a>|@tomprince commented|
|-|-|

(In [42375]) Branching to sslverify-no-pickle-6166-4.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_36"></a>|@tomprince set status to `closed`|
|-|-|

(In [42380]) Merge sslverify-no-pickle-6166-4: Deprecate pickling support in twisted.internet._sslverfiy.

Author: canibanoglu, tom.prince
Reviewer: exarkun, tom.prince, rwall
Fixes: [#6166](https://github.com/twisted/twisted/issues/6166)

Pickling twisted.internet.ssl.OptionSSLCertificationOptions and
twisted.internet.ssl.Keypair is no longer supported.
